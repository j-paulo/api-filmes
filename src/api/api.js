import axios from "axios";

const api = axios.create({
  baseURL: "https://swapi.dev/api/",
});

export const fetchFilmes = async (url = "films") => {
  const resposta = await api.get(url);

  return resposta.data.results;
};
