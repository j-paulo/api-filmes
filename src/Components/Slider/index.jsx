import React, { useState, useEffect } from "react";
import "react-responsive-carousel/lib/styles/carousel.min.css";
import Carousel from "react-responsive-carousel/lib/js/components/Carousel/index";
import Conteudo from "../Conteudo";
import imgCapa from "../../assets/images/capa.png";
import imgCapaMobile from "../../assets/images/capamobile.jpg";
import { BoxConteudo, Titulo, Texto, Botao, Image } from "../UI";
import { fetchFilmes } from "../../api/api";
import Loader from "../Loader/";
import { trackPromise, usePromiseTracker } from "react-promise-tracker";

export const ordencao = { changeOrdem: null };

const Slider = () => {
  const [selectedItem, setSelectedItem] = useState(0);
  const [filmes, setFilmes] = useState([]);
  const { promiseInProgress } = usePromiseTracker();

  ordencao.changeOrdem = (ordem) => {
    let filmesOrdenados = filmes.sort((a, b) => a.episode_id - b.episode_id);

    if (ordem === "lancamento") {
      const metadeFilmes = filmesOrdenados.length / 2;
      const primeirosFilmes = filmesOrdenados.slice(0, metadeFilmes);
      const ultimosFilmes = filmesOrdenados.slice(metadeFilmes);

      filmesOrdenados = [...ultimosFilmes, ...primeirosFilmes];
    }

    setFilmes([...filmesOrdenados]);
    setSelectedItem(selectedItem + 1);
  };

  useEffect(() => {
    let mounted = true;

    const getFilmes = async () => {
      const data = await fetchFilmes();

      if (mounted) setFilmes(data);
    };

    trackPromise(getFilmes());

    return () => {
      mounted = false;
    };
  }, []);

  return (
    <Carousel
      dynamicHeight
      showThumbs={false}
      showStatus={false}
      selectedItem={selectedItem}
      onChange={(index) => setSelectedItem(index)}
    >
      <div>
        <BoxConteudo>
          <Titulo>Filmoteca Star Wars</Titulo>
          <Texto>
            Seja bem vindo ao maior acervo de filmes da obra maxima de George
            Lucas!
          </Texto>
          <Texto>
            Aqui voce encontrara tudo sobre uma das maiores franquias do cinema!
          </Texto>
          <Texto>
            Vamos comecar?
            <br />
            Selecione abaixo a ordem em que deseja exibir os filmes:
          </Texto>
          {promiseInProgress === true ? (
            <Loader />
          ) : (
            <>
              <Botao
                onClick={() => ordencao.changeOrdem("cronologica")}
                style={{ marginRight: "25px" }}
              >
                - Ordem Cronologica -
              </Botao>
              <Botao onClick={() => ordencao.changeOrdem("lancamento")}>
                - Ordem de Lancamento -
              </Botao>
            </>
          )}
        </BoxConteudo>
        <Image className="desk" src={imgCapa} alt="" />
        <Image className="mobile" src={imgCapaMobile} alt="" />
      </div>
      {filmes.map((filme, key) => (
        <div key={key}>
          <Conteudo filme={filme} />
        </div>
      ))}
    </Carousel>
  );
};

export default Slider;
