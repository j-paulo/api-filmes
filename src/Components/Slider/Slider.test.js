import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import Slider, { ordencao } from ".";
import { fetchFilmes } from "../../api/api";

jest.mock("../../api/api");

const mockValues = [
  {
    episode_id: 1,
    title: "teste",
    opening_crawl: "star",
    director: "stevie",
    producer: "Loki",
    release_date: "1977-05-25",
  },
  {
    episode_id: 2,
    title: "new moon",
    opening_crawl: "another",
    director: "samuel",
    producer: "jojo",
    release_date: "1977-05-25",
  },
  {
    episode_id: 3,
    title: "imperio",
    opening_crawl: "de volta",
    director: "richard",
    producer: "alamo",
    release_date: "1977-05-25",
  },
];

describe("Componente de Slider", () => {
  describe("Quando o componente e montado", () => {
    it("Os botoes de ordenacao sao exibidos apos o conteudo da API ser carregado", async () => {
      fetchFilmes.mockResolvedValue(mockValues);

      render(<Slider />);

      const buttons = await screen.findAllByRole("button", {
        name: /ordem/i,
      });

      expect(buttons).toHaveLength(2);
    });
    it("A lista de filmes e carregada no carousel", async () => {
      fetchFilmes.mockResolvedValue(mockValues);

      const { container } = render(<Slider />);

      expect(await screen.findByText("teste")).toBeInTheDocument();

      expect(container.querySelector("ul.slider").children.length).toBe(4);
    });
  });
  describe("Quando eu clico no botao", () => {
    it("a funcao de ordenacao e executada", async () => {
      fetchFilmes.mockResolvedValue(mockValues);

      render(<Slider />);

      const button = await screen.findByRole("button", {
        name: /ordem cronologica/i,
      });

      const changeOrdem = jest.spyOn(ordencao, "changeOrdem");

      fireEvent.click(button);

      expect(changeOrdem).toHaveBeenCalled();
    });
    it("o carousel vai para o primeiro item", async () => {
      fetchFilmes.mockResolvedValue(mockValues);

      const { container } = render(<Slider />);

      const button = await screen.findByRole("button", {
        name: /ordem cronologica/i,
      });

      fireEvent.click(button);

      const itemContent = container.querySelector("li.slide.selected h1");

      expect(itemContent).toHaveTextContent("test");
    });

    describe("por ordem de lancamento", () => {
      it("os filmes sao ordenados por lancamento", async () => {
        fetchFilmes.mockResolvedValue(mockValues);

        const { container } = render(<Slider />);

        const button = await screen.findByRole("button", {
          name: /ordem de lancamento/i,
        });

        fireEvent.click(button);

        expect(container).toMatchSnapshot();
      });
    });
    describe("por ordem cronologica", () => {
      it("os filmes sao ordenados por ordem cronologica", async () => {
        fetchFilmes.mockResolvedValue(mockValues);

        const { container } = render(<Slider />);

        const button = await screen.findByRole("button", {
          name: /ordem cronologica/i,
        });

        fireEvent.click(button);

        expect(container).toMatchSnapshot();
      });
    });
  });
});
