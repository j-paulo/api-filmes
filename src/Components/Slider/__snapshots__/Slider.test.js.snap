// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Componente de Slider Quando eu clico no botao por ordem cronologica os filmes sao ordenados por ordem cronologica 1`] = `
<div>
  <div
    class="carousel-root"
  >
    <div
      class="carousel carousel-slider"
      style="width: 100%;"
    >
      <ul
        class="control-dots"
      >
        <li
          aria-label="slide item 1"
          class="dot"
          role="button"
          tabindex="0"
          value="0"
        />
        <li
          aria-label="slide item 2"
          class="dot selected"
          role="button"
          tabindex="0"
          value="1"
        />
        <li
          aria-label="slide item 3"
          class="dot"
          role="button"
          tabindex="0"
          value="2"
        />
        <li
          aria-label="slide item 4"
          class="dot"
          role="button"
          tabindex="0"
          value="3"
        />
      </ul>
      <button
        aria-label="previous slide / item"
        class="control-arrow control-prev"
        type="button"
      />
      <div
        class="slider-wrapper axis-horizontal"
        style="height: auto;"
      >
        <ul
          class="slider animated"
          style="transform: translate3d(-100%,0,0); transition-duration: 350ms;"
        >
          <li
            class="slide previous"
          >
            <div>
              <div
                class="sc-bdfBQB iqGwor"
              >
                <h1
                  class="sc-gsTEea bMRqBP"
                >
                  Filmoteca Star Wars
                </h1>
                <p
                  class="sc-dlfnuX cSnlMk"
                >
                  Seja bem vindo ao maior acervo de filmes da obra maxima de George Lucas!
                </p>
                <p
                  class="sc-dlfnuX cSnlMk"
                >
                  Aqui voce encontrara tudo sobre uma das maiores franquias do cinema!
                </p>
                <p
                  class="sc-dlfnuX cSnlMk"
                >
                  Vamos comecar?
                  <br />
                  Selecione abaixo a ordem em que deseja exibir os filmes:
                </p>
                <button
                  class="sc-hKgJUU hLjuup"
                  style="margin-right: 25px;"
                >
                  - Ordem Cronologica -
                </button>
                <button
                  class="sc-hKgJUU hLjuup"
                >
                  - Ordem de Lancamento -
                </button>
              </div>
              <img
                alt=""
                class="sc-eCstlR fbOgtV desk"
                src="capa.png"
              />
              <img
                alt=""
                class="sc-eCstlR fbOgtV mobile"
                src="capamobile.jpg"
              />
            </div>
          </li>
          <li
            class="slide selected"
          >
            <div>
              <div
                class="sc-bdfBQB iqGwor"
              >
                <h1
                  class="sc-gsTEea bMRqBP"
                >
                  teste
                </h1>
                <p
                  class="sc-dlfnuX cSnlMk"
                >
                  star
                </p>
                <button
                  class="sc-hKgJUU sc-iBPTik hLjuup yFKxa"
                >
                  - Mais infos -
                </button>
              </div>
              <div
                class="sc-bdfBQB sc-jSgvzq iqGwor kUZHzP"
              >
                <span
                  class="sc-gKseQn jmSvDS"
                >
                  Diretor: 
                  <font>
                    stevie
                  </font>
                </span>
                <br />
                <span
                  class="sc-gKseQn jmSvDS"
                >
                  Produtor: 
                  <font>
                    Loki
                  </font>
                </span>
                <br />
                <span
                  class="sc-gKseQn jmSvDS"
                >
                  Data de lancamento: 
                  <font>
                    24/05/1977
                  </font>
                </span>
                <br />
              </div>
              <img
                alt=""
                class="sc-eCstlR fbOgtV desk"
                src="phantommenace.png"
              />
              <img
                alt=""
                class="sc-eCstlR fbOgtV mobile"
                src="phantommenacemobile.jpg"
              />
            </div>
          </li>
          <li
            class="slide"
          >
            <div>
              <div
                class="sc-bdfBQB iqGwor"
              >
                <h1
                  class="sc-gsTEea bMRqBP"
                >
                  new moon
                </h1>
                <p
                  class="sc-dlfnuX cSnlMk"
                >
                  another
                </p>
                <button
                  class="sc-hKgJUU sc-iBPTik hLjuup yFKxa"
                >
                  - Mais infos -
                </button>
              </div>
              <div
                class="sc-bdfBQB sc-jSgvzq iqGwor kUZHzP"
              >
                <span
                  class="sc-gKseQn jmSvDS"
                >
                  Diretor: 
                  <font>
                    samuel
                  </font>
                </span>
                <br />
                <span
                  class="sc-gKseQn jmSvDS"
                >
                  Produtor: 
                  <font>
                    jojo
                  </font>
                </span>
                <br />
                <span
                  class="sc-gKseQn jmSvDS"
                >
                  Data de lancamento: 
                  <font>
                    24/05/1977
                  </font>
                </span>
                <br />
              </div>
              <img
                alt=""
                class="sc-eCstlR fbOgtV desk"
                src="attackclones.png"
              />
              <img
                alt=""
                class="sc-eCstlR fbOgtV mobile"
                src="attackclonesmobile.jpg"
              />
            </div>
          </li>
          <li
            class="slide"
          >
            <div>
              <div
                class="sc-bdfBQB iqGwor"
              >
                <h1
                  class="sc-gsTEea bMRqBP"
                >
                  imperio
                </h1>
                <p
                  class="sc-dlfnuX cSnlMk"
                >
                  de volta
                </p>
                <button
                  class="sc-hKgJUU sc-iBPTik hLjuup yFKxa"
                >
                  - Mais infos -
                </button>
              </div>
              <div
                class="sc-bdfBQB sc-jSgvzq iqGwor kUZHzP"
              >
                <span
                  class="sc-gKseQn jmSvDS"
                >
                  Diretor: 
                  <font>
                    richard
                  </font>
                </span>
                <br />
                <span
                  class="sc-gKseQn jmSvDS"
                >
                  Produtor: 
                  <font>
                    alamo
                  </font>
                </span>
                <br />
                <span
                  class="sc-gKseQn jmSvDS"
                >
                  Data de lancamento: 
                  <font>
                    24/05/1977
                  </font>
                </span>
                <br />
              </div>
              <img
                alt=""
                class="sc-eCstlR fbOgtV desk"
                src="revengesith.png"
              />
              <img
                alt=""
                class="sc-eCstlR fbOgtV mobile"
                src="revengesithmobile.jpg"
              />
            </div>
          </li>
        </ul>
      </div>
      <button
        aria-label="next slide / item"
        class="control-arrow control-next"
        type="button"
      />
    </div>
  </div>
</div>
`;

exports[`Componente de Slider Quando eu clico no botao por ordem de lancamento os filmes sao ordenados por lancamento 1`] = `
<div>
  <div
    class="carousel-root"
  >
    <div
      class="carousel carousel-slider"
      style="width: 100%;"
    >
      <ul
        class="control-dots"
      >
        <li
          aria-label="slide item 1"
          class="dot"
          role="button"
          tabindex="0"
          value="0"
        />
        <li
          aria-label="slide item 2"
          class="dot selected"
          role="button"
          tabindex="0"
          value="1"
        />
        <li
          aria-label="slide item 3"
          class="dot"
          role="button"
          tabindex="0"
          value="2"
        />
        <li
          aria-label="slide item 4"
          class="dot"
          role="button"
          tabindex="0"
          value="3"
        />
      </ul>
      <button
        aria-label="previous slide / item"
        class="control-arrow control-prev"
        type="button"
      />
      <div
        class="slider-wrapper axis-horizontal"
        style="height: auto;"
      >
        <ul
          class="slider animated"
          style="transform: translate3d(-100%,0,0); transition-duration: 350ms;"
        >
          <li
            class="slide previous"
          >
            <div>
              <div
                class="sc-bdfBQB iqGwor"
              >
                <h1
                  class="sc-gsTEea bMRqBP"
                >
                  Filmoteca Star Wars
                </h1>
                <p
                  class="sc-dlfnuX cSnlMk"
                >
                  Seja bem vindo ao maior acervo de filmes da obra maxima de George Lucas!
                </p>
                <p
                  class="sc-dlfnuX cSnlMk"
                >
                  Aqui voce encontrara tudo sobre uma das maiores franquias do cinema!
                </p>
                <p
                  class="sc-dlfnuX cSnlMk"
                >
                  Vamos comecar?
                  <br />
                  Selecione abaixo a ordem em que deseja exibir os filmes:
                </p>
                <button
                  class="sc-hKgJUU hLjuup"
                  style="margin-right: 25px;"
                >
                  - Ordem Cronologica -
                </button>
                <button
                  class="sc-hKgJUU hLjuup"
                >
                  - Ordem de Lancamento -
                </button>
              </div>
              <img
                alt=""
                class="sc-eCstlR fbOgtV desk"
                src="capa.png"
              />
              <img
                alt=""
                class="sc-eCstlR fbOgtV mobile"
                src="capamobile.jpg"
              />
            </div>
          </li>
          <li
            class="slide selected"
          >
            <div>
              <div
                class="sc-bdfBQB iqGwor"
              >
                <h1
                  class="sc-gsTEea bMRqBP"
                >
                  new moon
                </h1>
                <p
                  class="sc-dlfnuX cSnlMk"
                >
                  another
                </p>
                <button
                  class="sc-hKgJUU sc-iBPTik hLjuup yFKxa"
                >
                  - Mais infos -
                </button>
              </div>
              <div
                class="sc-bdfBQB sc-jSgvzq iqGwor kUZHzP"
              >
                <span
                  class="sc-gKseQn jmSvDS"
                >
                  Diretor: 
                  <font>
                    samuel
                  </font>
                </span>
                <br />
                <span
                  class="sc-gKseQn jmSvDS"
                >
                  Produtor: 
                  <font>
                    jojo
                  </font>
                </span>
                <br />
                <span
                  class="sc-gKseQn jmSvDS"
                >
                  Data de lancamento: 
                  <font>
                    24/05/1977
                  </font>
                </span>
                <br />
              </div>
              <img
                alt=""
                class="sc-eCstlR fbOgtV desk"
                src="attackclones.png"
              />
              <img
                alt=""
                class="sc-eCstlR fbOgtV mobile"
                src="attackclonesmobile.jpg"
              />
            </div>
          </li>
          <li
            class="slide"
          >
            <div>
              <div
                class="sc-bdfBQB iqGwor"
              >
                <h1
                  class="sc-gsTEea bMRqBP"
                >
                  imperio
                </h1>
                <p
                  class="sc-dlfnuX cSnlMk"
                >
                  de volta
                </p>
                <button
                  class="sc-hKgJUU sc-iBPTik hLjuup yFKxa"
                >
                  - Mais infos -
                </button>
              </div>
              <div
                class="sc-bdfBQB sc-jSgvzq iqGwor kUZHzP"
              >
                <span
                  class="sc-gKseQn jmSvDS"
                >
                  Diretor: 
                  <font>
                    richard
                  </font>
                </span>
                <br />
                <span
                  class="sc-gKseQn jmSvDS"
                >
                  Produtor: 
                  <font>
                    alamo
                  </font>
                </span>
                <br />
                <span
                  class="sc-gKseQn jmSvDS"
                >
                  Data de lancamento: 
                  <font>
                    24/05/1977
                  </font>
                </span>
                <br />
              </div>
              <img
                alt=""
                class="sc-eCstlR fbOgtV desk"
                src="revengesith.png"
              />
              <img
                alt=""
                class="sc-eCstlR fbOgtV mobile"
                src="revengesithmobile.jpg"
              />
            </div>
          </li>
          <li
            class="slide"
          >
            <div>
              <div
                class="sc-bdfBQB iqGwor"
              >
                <h1
                  class="sc-gsTEea bMRqBP"
                >
                  teste
                </h1>
                <p
                  class="sc-dlfnuX cSnlMk"
                >
                  star
                </p>
                <button
                  class="sc-hKgJUU sc-iBPTik hLjuup yFKxa"
                >
                  - Mais infos -
                </button>
              </div>
              <div
                class="sc-bdfBQB sc-jSgvzq iqGwor kUZHzP"
              >
                <span
                  class="sc-gKseQn jmSvDS"
                >
                  Diretor: 
                  <font>
                    stevie
                  </font>
                </span>
                <br />
                <span
                  class="sc-gKseQn jmSvDS"
                >
                  Produtor: 
                  <font>
                    Loki
                  </font>
                </span>
                <br />
                <span
                  class="sc-gKseQn jmSvDS"
                >
                  Data de lancamento: 
                  <font>
                    24/05/1977
                  </font>
                </span>
                <br />
              </div>
              <img
                alt=""
                class="sc-eCstlR fbOgtV desk"
                src="phantommenace.png"
              />
              <img
                alt=""
                class="sc-eCstlR fbOgtV mobile"
                src="phantommenacemobile.jpg"
              />
            </div>
          </li>
        </ul>
      </div>
      <button
        aria-label="next slide / item"
        class="control-arrow control-next"
        type="button"
      />
    </div>
  </div>
</div>
`;
